File: Fundamentals_of_Software_Design/1_Basic_skills/Выбывание из круга/main.py
def josephus_problem(n, k):
    # Список пользователей
    users = list(range(1, n + 1))
    index = 0  # Индекс для отсчета
    
    # Пока в списке больше одного пользователя
    while len(users) > 1:
        # Вычисляем индекс для выбывающего пользователя
        index = (index + k - 1) % len(users)
        # Выводим номер выбывающего
        print(users.pop(index))

    # Последний оставшийся пользователь
    print(users[0])

if __name__ == '__main__':
    n = int(input("Введите количество пользователей (N): "))
    k = int(input("Введите шаг выбывания (K): "))
    
    josephus_problem(n, k)

File: Fundamentals_of_Software_Design/1_Basic_skills/Выбывание из круга/main.c
#include <stdio.h>
#include <stdlib.h>

void josephus_problem(int n, int k) {
    // Создаем массив с пользователями
    int *users = (int *)malloc(n * sizeof(int));
    
    // Заполняем массив номерами пользователей
    for (int i = 0; i < n; i++) {
        users[i] = i + 1;
    }

    int index = 0;  // Индекс для отсчета
    int remaining = n;  // Количество оставшихся пользователей

    // Пока не остался один пользователь
    while (remaining > 1) {
        // Находим индекс выбывающего пользователя
        index = (index + k - 1) % remaining;

        // Выводим номер выбывающего
        printf("%d\n", users[index]);

        // Сдвигаем элементы, чтобы удалить выбывшего
        for (int i = index; i < remaining - 1; i++) {
            users[i] = users[i + 1];
        }

        remaining--;  // Уменьшаем количество оставшихся
    }

    // Последний оставшийся пользователь
    printf("%d\n", users[0]);

    free(users);  // Освобождаем память
}

int main() {
    int n, k;
    
    // Вводим количество пользователей и шаг выбывания
    printf("Введите количество пользователей (N): ");
    scanf("%d", &n);
    printf("Введите шаг выбывания (K): ");
    scanf("%d", &k);
    
    josephus_problem(n, k);

    return 0;
}
